// ag-grid-angular/esm2022/lib/angularFrameworkEventListenerService.mjs
export class AngularFrameworkEventListenerService {
    constructor(frameworkOverrides) {
        this.frameworkOverrides = frameworkOverrides;
        // Map from user listener to wrapped listener so we can remove listener provided by user
        this.wrappedListeners = new Map();
        this.wrappedGlobalListeners = new Map();
    }
    wrap(eventType, userListener) {
        const { frameworkOverrides, wrappedListeners } = this;
        let listener = userListener;
        if (frameworkOverrides.shouldWrapOutgoing) {
            listener = (event) => {
                frameworkOverrides.wrapOutgoing(() => userListener(event));
            };
            let eventListeners = wrappedListeners.get(eventType);
            if (!eventListeners) {
                eventListeners = new Map();
                wrappedListeners.set(eventType, eventListeners);
            }
            eventListeners.set(userListener, listener);
        }
        return listener;
    }
    wrapGlobal(userListener) {
        const { frameworkOverrides, wrappedGlobalListeners } = this;
        let listener = userListener;
        if (frameworkOverrides.shouldWrapOutgoing) {
            listener = (eventType, event) => {
                frameworkOverrides.wrapOutgoing(() => userListener(eventType, event));
            };
            wrappedGlobalListeners.set(userListener, listener);
        }
        return listener;
    }
    unwrap(eventType, userListener) {
        const { wrappedListeners } = this;
        const eventListeners = wrappedListeners.get(eventType);
        if (eventListeners) {
            const wrapped = eventListeners.get(userListener);
            if (wrapped) {
                eventListeners.delete(userListener);
                if (eventListeners.size === 0) {
                    wrappedListeners.delete(eventType);
                }
                return wrapped;
            }
        }
        return userListener;
    }
    unwrapGlobal(userListener) {
        const { wrappedGlobalListeners } = this;
        const wrapped = wrappedGlobalListeners.get(userListener);
        if (wrapped) {
            wrappedGlobalListeners.delete(userListener);
            return wrapped;
        }
        return userListener;
    }
}
